drivers: add infrastructure for managing device MMIO regions

tyrantlizard77 opened the pull request:
PCI-E, and the implementation of virtual memory imply that the linear address for accessing driver MMIO regions is not necessarily known at build time. So far, drivers have only implemented ad hoc solutions to managing where this data gets stored, or the region address is just hard-coded in ROM or RAM.

This patch:

- Introduces a header with macros for managing driver I/O regions, appropriately storing data in either RAM or ROM depending on configuration.
- Introduces a new Kconfig CONFIG_MMU, `k_mem_map()` and `arch_mem_map()` APIs for MMU-based systems, with an implementation for x86. For this PR there's just enough here to identity map driver regions, more to come in future PRs as I make more progress on https://github.com/zephyrproject-rtos/zephyr/issues/26088. At the moment this is more or less just a stub.
- Updates various drivers used on x86 to use these macros

This mechanism should have no negative effect on footprint for systems without an MMU or PCI-E.

Fixes: #26168
Fixes: #26107

Commit:
arch: add CONFIG_CPU_HAS_MMU

Indicate that the CPU has a memory management unit,
similar to CPU_HAS_MPU for MPUs.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arch: add CONFIG_MMU

This config indicates that a memory management unit is present
and enabled, which will in turn allow arch APIs to allow
mapping memory to be used.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add macros for device MMIO management

Three variants provided:

- Drivers with one MMIO region to manage
- Drivers with multiple MMIO regions to manage
- Drivers or driver-like code which doesn't use Zephyr's device
  model

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: early_serial: use device_map()

This driver code uses PCIe and doesn't use Zephyr's
device model, so we can't use the nice DEVICE_MMIO macros.
Set stuff up manually instead using device_map().

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: eth_e1000: use device MMIO

The ethernet driver no longer continues to try to initialize
itself if PCIe probing fails.

This device is always PCIe so we don't need to reserve ROM MMIO
storage.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: apic: use device MMIO APIs

A hack was required for the loapic code due to the address
range not being in DTS. A bug was filed.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: hpet: use device_mmio

Straightforward example on how to use the TOPLEVEL variants
of the DEVICE_MMIO macros.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add k_mem_map() and related defines

This will be the interface for mapping memory in the kernel's
part of the address space, which is guaranteed to be persistent
regardless of what thread is scheduled.

Further code for specifically managing virtual memory will end up in
kernel/mmu.c.

Further defintions for memory management in general will end up
in sys/mem_manage.h.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: define arch_mem_map()

This is the low-level arch function to map a region into page
tables.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: i2c_dw: use device_mmio

Fairly straightforward example of how to deal with PCIe.
The code considers whether a particular instance is PCIe
or not on a per-instance basis, so DEVICE_MMIO_ROM is not
conditionally defined.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
gpio_intel_apl: use DEVICE_MMIO

Good example of the NAMED variants of the MMIO macros, since
an existing inheritance mechanism already took the first-member
slot of the dev_data/config structs.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: pcie: use device_map() for MMIO config access

Replaces custom runtime calls to map memory.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: remove memory mapping SOC code

This isn't needed any more, all of these directives were
for drivers which use device_map() now.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: add arch_mem_map()

This currently only supports identity paging; there's just
enough here for device_map() calls to work.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
doc: add device MMIO details

Add some overview documentation on how to use this header.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
The documentation errors make no sense to my eye, not sure where these `<em>` `</em>` tags are coming from

```
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:437: warning: argument 'flags' of command @param is not found in the argument list of DEVICE_MMIO_NAMED_MAP(_device, _name, _flags)
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:458: warning: found </em> tag without matching <em>
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:463: warning: The following parameters of DEVICE_MMIO_NAMED_MAP(_device, _name, _flags) are not documented:
  parameter '_flags'
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:597: warning: argument 'em' of command @param is not found in the argument list of DEVICE_MMIO_TOPLEVEL_MAP(_name, _flags)
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:597: warning: argument 'flags' of command @param is not found in the argument list of DEVICE_MMIO_TOPLEVEL_MAP(_name, _flags)
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:611: warning: found </em> tag without matching <em>
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:616: warning: The following parameters of DEVICE_MMIO_TOPLEVEL_MAP(_name, _flags) are not documented:
  parameter '_flags'
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:437: warning: argument 'em' of command @param is not found in the argument list of DEVICE_MMIO_NAMED_MAP(_device, _name, _flags)
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:437: warning: argument 'flags' of command @param is not found in the argument list of DEVICE_MMIO_NAMED_MAP(_device, _name, _flags)
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:458: warning: found </em> tag without matching <em>
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:463: warning: The following parameters of DEVICE_MMIO_NAMED_MAP(_device, _name, _flags) are not documented:
  parameter '_flags'
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:597: warning: argument 'em' of command @param is not found in the argument list of DEVICE_MMIO_TOPLEVEL_MAP(_name, _flags)
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:597: warning: argument 'flags' of command @param is not found in the argument list of DEVICE_MMIO_TOPLEVEL_MAP(_name, _flags)
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:611: warning: found </em> tag without matching <em>
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:616: warning: The following parameters of DEVICE_MMIO_TOPLEVEL_MAP(_name, _flags) are not documented:
  parameter '_flags'
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:437: warning: argument 'em' of command @param is not found in the argument list of DEVICE_MMIO_NAMED_MAP(_device, _name, _flags)
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:437: warning: argument 'flags' of command @param is not found in the argument list of DEVICE_MMIO_NAMED_MAP(_device, _name, _flags)
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:458: warning: found </em> tag without matching <em>
/home/runner/work/zephyr/zephyr/include/sys/device_mmio.h:463: warning: The following parameters of DEVICE_MMIO_NAMED_MAP(_device, _name, _flags) are not documented:
  parameter '_flags'
```

tbursztyka reviewed (approved):
neat

jhedberg commented on include/sys/mm.h:
```diff
@@ -0,0 +1,109 @@
+extern "C" {
+#endif
+
+#ifdef CONFIG_MMU
```
I think it'd be better *not* to have such #ifdefs in header files for function prototypes, since it prevents using `IS_ENABLED(CONFIG_MMU)` in the calling code. I.e. leave it up to the linker to throw the error of a non-existing function rather than the compiler. An additional benefit is that you don't need to add exceptions to the documentation building to make sure these get included included there.

andrewboie replied:
OK, removed.

jhedberg reviewed (changes requested):
Looks quite good, minor comment regarding #ifdef usage in header files

pabigot commented on include/sys/mm.h:
```diff
@@ -0,0 +1,109 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_MM_H
```
`mm` is a little opaque; could this be `<sys/memmap.h>` instead? Or `<sys/memmgmt.h>` if that's a better name?

tbursztyka replied:
I'd vote for memmap.h

andrewboie replied:
I renamed it to `mem_manage.h`.

pabigot reviewed (dismissed):
One request to use an include file name that's more descriptive.  The rest seems impressively completely and well documented, but I'm not qualified to approve it.

Commit:
tests: device: add device_mmio tests

Exercise the public macros as well as device_map().
This test has a whitelist for whatever reason; add
mps2_an385 so that the !DEVICE_MMIO_IS_IN_RAM stuff
is tested.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: mslab_threadsafe: cleanup

Clean up the code some, and disable CONFIG_TIMESLICE_SIZE=1
until #13813 is resolved.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: ns16550: use DEVICE_MMIO where needed

This gets a little complicated as the driver could be
using PCI-E MMIO, MMIO specified by DTS, or I/O ports.
This driver doesn't use struct uart_device_config any
more.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
I updated the branch and took a much simpler approach to converting ns16550 and the PR no longer touches other serial drivers or modifies the subsystem header. I also added a workaround for https://github.com/zephyrproject-rtos/zephyr/issues/13813 which started popping up again with this change (consistently, due to icount).

I seem to be having a problem with GitHub, when I push new changes to my branch the PR is not being updated, even though I can see the new changes if I browse the branch manually here: https://github.com/andrewboie/zephyr/tree/device-mmio

If I can't figure out how to get GH to stop malfunctioning I may need to close this and submit a new PR, I'd really rather not do that though.
